File: packages/host-node/src/process/environment.ts
/**
 * The environment every Git process is given.
 *
 * Two problems are solved here, and both have bitten real tools:
 *
 * 1. **Ambient redirects.** A `GIT_DIR`, `GIT_WORK_TREE`, `GIT_INDEX_FILE`,
 *    `GIT_OBJECT_DIRECTORY` or `GIT_CONFIG` in the caller's shell would silently
 *    point Git at a different repository, index or configuration than the one the
 *    user approved. They are removed, not inherited.
 * 2. **Interactive Git.** `GIT_TERMINAL_PROMPT=0` stops Git waiting on a terminal
 *    for a password that will never come. That does not make every helper
 *    non-interactive — an SSH askpass or a credential helper can still prompt — so
 *    the design's rule applies: report that credentials are needed on the target
 *    machine rather than disabling host verification to make it go away.
 *
 * Everything else the user's environment provides is kept, because this service
 * runs as that user against *their* repositories: `HOME` finds their global
 * config, `SSH_AUTH_SOCK` their agent, `GPG_TTY`/`GNUPGHOME` their signing setup.
 * A client cannot influence any of this — the environment is built from the host
 * process, never from a request.
 */
import { delimiter, dirname } from "node:path";

/** Variables copied from the host process when present, and only these. */
export const INHERITED_ENV_VARS: readonly string[] = [
  // Executable and shell basics.
  "PATH",
  "PATHEXT",
  "SHELL",
  "COMSPEC",
  // The user, whose repositories and credentials these are.
  "HOME",
  "USER",
  "LOGNAME",
  "USERPROFILE",
  "HOMEDRIVE",
  "HOMEPATH",
  "APPDATA",
  "LOCALAPPDATA",
  "XDG_CONFIG_HOME",
  "XDG_CACHE_HOME",
  "XDG_DATA_HOME",
  // The commit identity, which belongs to whoever started this service. These set a
  // name, an address and a date — they cannot redirect Git or make it run anything, so
  // they are not on the blocked list — and dropping them would mean two visible
  // failures: a CI job that exported GIT_COMMITTER_* getting commits attributed to
  // somebody else, and Git resolving an identity from the system account database for
  // every command that needs one. That fallback was measured at 15.05s for a single
  // `git worktree add` on the machine this was found on.
  "GIT_AUTHOR_NAME",
  "GIT_AUTHOR_EMAIL",
  "GIT_AUTHOR_DATE",
  "GIT_COMMITTER_NAME",
  "GIT_COMMITTER_EMAIL",
  "GIT_COMMITTER_DATE",
  // Signing and SSH, so the user's existing setup keeps working.
  "SSH_AUTH_SOCK",
  "SSH_AGENT_PID",
  "GNUPGHOME",
  "GPG_TTY",
  "TERM",
  "DISPLAY",
  // Temp space and platform paths.
  "TMPDIR",
  "TEMP",
  "TMP",
  "SYSTEMROOT",
  "SYSTEMDRIVE",
  "WINDIR",
  // Locale and timezone, left as the user set them: Git's machine formats are
  // locale-independent, and diagnostics read better in the user's language.
  "LANG",
  "LC_ALL",
  "LC_MESSAGES",
  "TZ",
];

/**
 * Variables that redirect Git, or make it run something, and are therefore never
 * passed through — even if a future edit adds one to the inherit list above.
 *
 * `GIT_CONFIG_PARAMETERS` and `GIT_CONFIG_COUNT`/`KEY_n`/`VALUE_n` are the
 * environment equivalent of `-c` and can set any config key, including the ones
 * that name a program to execute (a diff driver, a textconv filter, a credential
 * helper, `core.sshCommand`).
 */
export const BLOCKED_ENV_VARS: readonly string[] = [
  "GIT_DIR",
  "GIT_WORK_TREE",
  "GIT_INDEX_FILE",
  "GIT_OBJECT_DIRECTORY",
  "GIT_ALTERNATE_OBJECT_DIRECTORIES",
  "GIT_COMMON_DIR",
  "GIT_NAMESPACE",
  "GIT_PREFIX",
  "GIT_CEILING_DIRECTORIES",
  "GIT_DISCOVERY_ACROSS_FILESYSTEM",
  "GIT_CONFIG",
  "GIT_CONFIG_COUNT",
  "GIT_CONFIG_PARAMETERS",
  "GIT_EXTERNAL_DIFF",
  "GIT_DIFF_OPTS",
  "GIT_TRACE",
  "GIT_TRACE2",
  "GIT_TRACE2_EVENT",
  "GIT_TRACE_PERFORMANCE",
  "GIT_TRACE_PACKET",
  "GIT_TRACE_SETUP",
  "GIT_TRACE_SHALLOW",
  "GIT_SSH_VARIANT",
];

/** Variables whose value is set by this host, never inherited. */
export const HOST_SET_ENV_VARS: Readonly<Record<string, string>> = {
  /** Never wait for a terminal prompt: there is no terminal. */
  GIT_TERMINAL_PROMPT: "0",
  /** No pager can take over a captured pipe; `--no-pager` is belt and braces. */
  GIT_PAGER: "cat",
  PAGER: "cat",
};

export interface BuildGitEnvironmentOptions {
  /** Absolute path of the Git executable, so its directory can lead PATH. */
  readonly gitPath?: string;
  /** Extra entries for tests; applied after the allow-list, minus the blocked names. */
  readonly extra?: Readonly<Record<string, string>>;
  /** The ambient environment; defaults to the host process's. */
  readonly source?: NodeJS.ProcessEnv;
}

/**
 * Build the child environment for one Git process.
 *
 * The result contains only inherited-and-allowed variables, the host-set values,
 * and (for tests) explicit extras. `gitPath`'s directory is prepended to PATH so
 * that a hook or a helper invoked by Git finds the same Git the service uses,
 * rather than whatever a modified PATH would offer. The blocked names are removed
 * last and cannot be reintroduced, not even by `extra`.
 */
export function buildGitEnvironment(
  options: BuildGitEnvironmentOptions = {},
): Record<string, string> {
  const source = options.source ?? process.env;
  const env: Record<string, string> = {};

  for (const name of INHERITED_ENV_VARS) {
    const value = source[name];
    if (value !== undefined && value.length > 0) {
      env[name] = value;
    }
  }

  if (options.gitPath !== undefined) {
    const directory = dirname(options.gitPath);
    const existing = env["PATH"];
    env["PATH"] =
      existing === undefined
        ? directory
        : `${directory}${delimiter}${existing}`;
  }

  for (const [name, value] of Object.entries(HOST_SET_ENV_VARS)) {
    env[name] = value;
  }
  if (options.extra !== undefined) {
    for (const [name, value] of Object.entries(options.extra)) {
      env[name] = value;
    }
  }

  // Blocked names are removed last, after every other source has been merged. The
  // block is structural rather than a default: `GIT_DIR`/`GIT_WORK_TREE` retarget
  // which repository a command operates on, and no caller — including a test's
  // `extra` — gets to put one back and make a scoped command act elsewhere.
  for (const name of BLOCKED_ENV_VARS) {
    delete env[name];
  }

  return env;
}

/** Names of the environment variables this host refuses to pass through. */
export function blockedVariableNames(): readonly string[] {
  return BLOCKED_ENV_VARS;
}

File: crates/refyard-host/src/process/environment.rs
//! The environment a Git command runs in.
//!
//! Two failure modes make this a module rather than a convenience:
//!
//! - **Inheriting everything is a correctness bug.** `GIT_DIR`, `GIT_WORK_TREE`,
//!   `GIT_INDEX_FILE` and `GIT_COMMON_DIR` redirect a command at a different
//!   repository than the one the user approved; `GIT_EXTERNAL_DIFF`, `GIT_SSH_COMMAND`
//!   and `GIT_ASKPASS` replace the program Git runs with one we did not choose. A
//!   developer machine and a CI runner both have such variables set often enough that
//!   "it worked on my machine" would otherwise be a real class of bug.
//! - **Inheriting nothing is also a bug.** Git needs `PATH`, and SSH needs the agent
//!   socket; a command started without them fails in a way that looks like a missing
//!   credential.
//!
//! So the rule is an allow-list plus forced values, and the list is here in one place
//! so it can be read and tested rather than inferred from a call site.

/// Variables passed through from the host process when present.
///
/// `SSH_AUTH_SOCK`/`SSH_AGENT_PID` are here on purpose: the user's key agent is the
/// intended way for SSH to authenticate, and hiding it would break the one
/// credential path this product supports.
pub const INHERITED: &[&str] = &[
    "PATH",
    "HOME",
    "USER",
    "LOGNAME",
    "SHELL",
    "TMPDIR",
    "TEMP",
    "TMP",
    "SystemRoot",
    "ComSpec",
    "LANG",
    "LC_ALL",
    "LC_CTYPE",
    "TERM",
    "TZ",
    "SSH_AUTH_SOCK",
    "SSH_AGENT_PID",
];

/// Values the host sets itself, overriding anything the environment claimed.
///
/// `GIT_TERMINAL_PROMPT=0` is not a preference: a command that stops to ask for a
/// password on a pipe nobody is reading would sit there until its deadline, and the
/// product has no way to answer it. Pagers are pinned because a pager would consume
/// the output the parser needs.
pub const FORCED: &[(&str, &str)] = &[
    ("GIT_TERMINAL_PROMPT", "0"),
    ("GIT_PAGER", "cat"),
    ("PAGER", "cat"),
];

/// Builds the environment for one run: only the allowed names, then the forced values.
///
/// The source is passed in rather than read from the process, so a test can state the
/// dangerous input explicitly and so a provider can substitute another environment
/// (a fixture's own `HOME`, for instance) without changing this rule.
pub fn sanitized_environment(source: &[(String, String)]) -> Vec<(String, String)> {
    let mut result: Vec<(String, String)> = Vec::new();
    for (name, value) in source {
        if INHERITED.iter().any(|allowed| allowed == name) {
            result.push((name.clone(), value.clone()));
        }
    }
    for (name, value) in FORCED {
        match result.iter_mut().find(|(existing, _)| existing == name) {
            Some(entry) => entry.1 = (*value).to_string(),
            None => result.push(((*name).to_string(), (*value).to_string())),
        }
    }
    result
}

/// The same rule, applied to this process's own environment.
pub fn git_environment() -> Vec<(String, String)> {
    let source: Vec<(String, String)> = std::env::vars().collect();
    sanitized_environment(&source)
}
